### PR TITLE
METRON-1168: Add SUBSTRING method to stellar

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments

Add `SUBSTRING` function to stellar.

Testing:
Open up the REPL and play with `SUBSTRING`

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [ ] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java [hunk 1]
```diff
@@ -408,6 +408,19 @@ public void testAppendIfMissing() throws Exception {
 
   }
 
+  @Test
+  public void testSubstring() throws Exception {
+    Map<String, Object> variables = ImmutableMap.of("s", "apache metron");
+    Assert.assertEquals("metron", run("SUBSTRING(s, 7)", variables));
+    Assert.assertEquals("me", run("SUBSTRING(s, 7, 9)", variables));
+    Assert.assertNull(run("SUBSTRING(null, 6, 9)", new HashMap<>()));
```
Comment (justinleet) [inline metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java]: Can you add a test for start being explicitly null, and another one for end being explicitly null?  I mostly care about start being null, since that's a case not present in Java, given that int can't be null.

**Comment by justinleet** [comment]
+1, thanks a lot for the contribution.  And for catching that we never actually had this

File: metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java (hunk 1 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java]: yep

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java [hunk 2]
```diff
@@ -321,6 +321,46 @@ public Object apply(List<Object> args) {
...
+                     }
+          , returns = "The substring of the input"
+  )
+  public static class Substring extends BaseStellarFunction {
+
+    @Override
+    public Object apply(List<Object> strings) {
+
+      if(strings == null || strings.size() < 2 ) {
+        throw new IllegalArgumentException("[SUBSTRING] required 2 arguments: the input and the start position (inclusive)");
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: The [] is not consistent with other functions #nit

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java [hunk 3]
```diff
@@ -321,6 +321,46 @@ public Object apply(List<Object> args) {
...
+      }
+      String var = strings.get(0) == null?null: (String) strings.get(0);
+      Integer start = strings.get(1) == null?null:(Integer)strings.get(1);
+      Integer end = null;
+      if(strings.size() > 2) {
+         end = strings.get(2) == null ? null : (Integer) strings.get(2);
+      }
+      if(var == null || start == null) {
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: We are again looking at the behavior with regards to invalid arguments, exception vs default return....   We need some clarification document.  It seems to come up every addition to Stellar

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java [hunk 4]
```diff
@@ -321,6 +321,46 @@ public Object apply(List<Object> args) {
...
+  public static class Substring extends BaseStellarFunction {
+
+    @Override
+    public Object apply(List<Object> strings) {
+
+      if(strings == null || strings.size() < 2 ) {
+        throw new IllegalArgumentException("[SUBSTRING] required 2 arguments: the input and the start position (inclusive)");
+      }
+      String var = strings.get(0) == null?null: (String) strings.get(0);
+      Integer start = strings.get(1) == null?null:(Integer)strings.get(1);
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: Type validation?

File: metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java [hunk 5]
```diff
@@ -408,6 +408,23 @@ public void testAppendIfMissing() throws Exception {
 
   }
 
+  @Test
+  public void testSubstring() throws Exception {
+    Map<String, Object> variables = ImmutableMap.of("s", "apache metron");
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java]: To go with type validation, tests with passing down incorrect types.

like : SUBSTRING(s,'7',9)

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java (hunk 2 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: Yeah, I was following the leader with `CHOMP`, but I agree, I will correct this in a follow-on JIRA with the rest of your issues covered.

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java (hunk 4 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: Agreed

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java (hunk 3 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/StringFunctions.java]: Well, this is consistent throughout stellar functions, if the value is null, we pass back null.  If it's invalid input structurally (i.e. not the correct number of arguments), then an exception is returned.  We should add a little blurb that talks about this in the stellar-common README.md

File: metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java (hunk 5 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/StringFunctionsTest.java]: Agreed, we should expect an exception in that case.

**Comment by cestella** [comment]
I am going to create a follow-on PR to this to address your issues.

**Comment by ottobackwards** [comment]
sounds good to me.  Thanks for the contribution!  I'm jealous I didn't think of it first.
